azure/services/privateendpoints/spec.go

```diff
@@ -54,209 +51,122 @@ type PrivateEndpointSpec struct {
 	ClusterName                   string
 }
 
-// ResourceName returns the name of the private endpoint.
-func (s *PrivateEndpointSpec) ResourceName() string {
-	return s.Name
-}
-
-// ResourceGroupName returns the name of the resource group.
-func (s *PrivateEndpointSpec) ResourceGroupName() string {
-	return s.ResourceGroup
-}
-
-// OwnerResourceName is a no-op for private endpoints.
-func (s *PrivateEndpointSpec) OwnerResourceName() string {
-	return ""
-}
-
-// Parameters returns the parameters for the PrivateEndpointSpec.
-func (s *PrivateEndpointSpec) Parameters(ctx context.Context, existing interface{}) (interface{}, error) {
-	_, log, done := tele.StartSpanWithLogger(ctx, "privateendpoints.Service.Parameters")
-	defer done()
-
-	privateEndpointProperties := armnetwork.PrivateEndpointProperties{
-		Subnet: &armnetwork.Subnet{
-			ID: &s.SubnetID,
-			Properties: &armnetwork.SubnetPropertiesFormat{
-				PrivateEndpointNetworkPolicies:    ptr.To(armnetwork.VirtualNetworkPrivateEndpointNetworkPoliciesDisabled),
-				PrivateLinkServiceNetworkPolicies: ptr.To(armnetwork.VirtualNetworkPrivateLinkServiceNetworkPoliciesEnabled),
-			},
+// ResourceRef implements azure.ASOResourceSpecGetter.
+func (s *PrivateEndpointSpec) ResourceRef() *asonetworkv1.PrivateEndpoint {
+	return &asonetworkv1.PrivateEndpoint{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      s.Name,
+			Namespace: s.Namespace,
 		},
 	}
+}
 
-	privateEndpointProperties.CustomNetworkInterfaceName = ptr.To(s.CustomNetworkInterfaceName)
-
-	privateIPAddresses := make([]*armnetwork.PrivateEndpointIPConfiguration, 0, len(s.PrivateIPAddresses))
-	for _, address := range s.PrivateIPAddresses {
-		ipConfig := &armnetwork.PrivateEndpointIPConfigurationProperties{PrivateIPAddress: ptr.To(address)}
-
-		privateIPAddresses = append(privateIPAddresses, &armnetwork.PrivateEndpointIPConfiguration{
-			Properties: ipConfig,
+// Parameters implements azure.ASOResourceSpecGetter.
```

Comment:
The `Parameters()` func fills up the fields of `asonetworkv1.PrivateEndpoint_Spec{}` only if they are present. 
This means that 
- _arrays_ with `length <= 0` do not get added.
- empty strings (`""` ) do not get added.

The reason to point this out specifically is that, in this implementation,
1. **Reason:**`Parameters()` func mimics the UX of a user creating an `asonetworkv1.PrivateEndpoint{}` type. Users will not care to populate nil values to the fields they are not interested in; they will only add the fields they are interested in. Therefore the untouched fields will default to their `nil types` and hence get omitted by the CRDs.
2. **Question:** Now that Parameters is mimicking UX of a user creating an `asonetworkv1.PrivateEndpoint{}` type , we may run into issues in the future if ever ASO decides to have different behavior for `nil types` and `empty types`. For instance, if ASO decides to have a different behavior for `privateEndpoint.Spec.Location = ptr.To("")` and `privateEndpoint.Spec.Location = nil`, CAPZ will run into errors. 
  - One way to solve this issue would be to update `s.Location` to be of type `*string` instead of `string`. But this could be a breaking change.

Reply:
I'd default to sticking to whatever we were doing before with the SDK for each field to reduce the risk of introducing regressions. From what I've seen, the ASO and SDK types match pretty closely in terms of what is and isn't a pointer.

I think trying to align that behavior across all the fields for every resource is out of scope for these PRs for now.

Reply:
> I think trying to align that behavior across all the fields for every resource is out of scope for these PRs for now.

I agree for the sake of exploding this PR with a ton of changes to review and merge. 

> the ASO and SDK types match pretty closely in terms of what is and isn't a pointer.

To a certain extent, yes, but not everything. 
For instance, `asonetworkv1.privateendpoint.PrivateEndpoint_Spec.Location` is a type `*string` and `PrivateEndpointSpec.Location` (at CAPZ side) is of type `string`. Although `Location` as a field is innocuous in this scenario; meaning an `""` and a `nil` `asonetworkv1.privateendpoint.PrivateEndpoint_Spec.Location` mean the same for the ASO CRD and will get dropped(due to omit empty annotation), I am suggesting that we might have to be wary in carrying over other fields at other places. 

Happy to revert to direct copy over of values btw!

Reply:
> But we're not changing the CAPZ type for Location here right?

Yes, we are not changing it. 

> So if we were translating a string from CAPZ into a *string for the SDK, can we do the same translation to *string for ASO?

I am not sure if we can do the same ? 

The way I understand it: 
- Working with SDK meant we created underlying Azure resources. 
- Working with ASO means we create a CR representing ASO resources's CRD, just like any user would manually create a CR. It is ASO's responsibility to create the Azure Resource for the requested CR.

So the fact that ASO chose to broadcast `asonetworkv1.privateendpoint.PrivateEndpoint_Spec.Location` as a type `*string` rather than a `string` suggests to me that they want to differentiate between a user initializing `asonetworkv1.privateendpoint.PrivateEndpoint_Spec.Location` and `""` value it gets assigned if left untouched.

(Although they list `Location` to be _omitted when empty_ (`Location *string 'json:"location,omitempty"'`) which means both `nil` and `""` would be treated the same way. 


I wonder if we will run into issues for fields of other data types like custom structs, etc.

Reply:
> The way I understand it:
> 
> * Working with SDK meant we created underlying Azure resources.
> * Working with ASO means we create a CR representing ASO resources's CRD, just like any user would manually create a CR. It is ASO's responsibility to create the Azure Resource for the requested CR.

These still seem like fundamentally the same thing to me though since ASO objects are basically just Kubernetes-shaped versions of the SDK objects.


> So the fact that ASO chose to broadcast `asonetworkv1.privateendpoint.PrivateEndpoint_Spec.Location` as a type `*string` rather than a `string` suggests to me that they want to differentiate between a user initializing `asonetworkv1.privateendpoint.PrivateEndpoint_Spec.Location` and `""` value it gets assigned if left untouched.

Right, and this was also the case with the SDK before, so I think we can keep the same behavior that we had previously.

> (Although they list `Location` to be _omitted when empty_ (`Location *string 'json:"location,omitempty"'`) which means both `nil` and `""` would be treated the same way.

With `omitempty` and any pointer type, only `nil` will get dropped from the serialized representation. The non-nil empty string is not omitted.

Reply:
TIL: https://goplay.tools/snippet/y3p2j_hB24w

Reply:
> > So the fact that ASO chose to broadcast `asonetworkv1.privateendpoint.PrivateEndpoint_Spec.Location` as a type `*string` rather than a `string` suggests to me that they want to differentiate between a user initializing `asonetworkv1.privateendpoint.PrivateEndpoint_Spec.Location` and `""` value it gets assigned if left untouched.
> 
> Right, and this was also the case with the SDK before, so I think we can keep the same behavior that we had previously.

👍🏼